cpp/include/cugraph/prims/gather_edges.cuh

```diff
@@ -0,0 +1,495 @@
+  std::exclusive_scan(
+    external_source_counts.begin(), external_source_counts.end(), displacements.begin(), size_t{0});
+
+  rmm::device_uvector<vertex_t> active_sources(total_external_source_count, handle.get_stream());
+  rmm::device_uvector<gpu_t> active_source_gpu_ids(total_external_source_count,
+                                                   handle.get_stream());
+  // Get the sources other gpus on the same row are working on
+  // TODO : replace with device_bcast for better scaling
+  device_allgatherv(col_comm,
+                    vertex_input_first,
+                    active_sources.data(),
+                    external_source_counts,
+                    displacements,
+                    handle.get_stream());
+  device_allgatherv(col_comm,
+                    gpu_id_first,
+                    active_source_gpu_ids.data(),
+                    external_source_counts,
+                    displacements,
+                    handle.get_stream());
+  thrust::sort_by_key(handle.get_thrust_policy(),
+                      active_sources.begin(),
+                      active_sources.end(),
+                      active_source_gpu_ids.begin());
+  return std::make_tuple(std::move(active_sources), std::move(active_source_gpu_ids));
+}
+
+/**
+ * @brief Gather valid edges present on the current gpu
+ *
+ * Collect all the edges that are present in the adjacency lists on the current gpu
+ *
+ * @tparam GraphViewType Type of the passed non-owning graph object.
+ * @tparam EdgeIndexIterator Type of the iterator for edge indices.
+ * @tparam GPUIdIterator  Type of the iterator for gpu id identifiers.
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param graph_view Non-owning graph object.
+ * @param active_sources_in_row Device vector containing all the vertex id that are processed by
+ * gpus in the column communicator
+ * @param active_source_gpu_ids Device vector containing the gpu id associated by every vertex
+ * present in active_sources_in_row
+ * @param edge_index_first Iterator pointing to the first destination index
+ * @param invalid_vertex_id Vertex id to fill in result if destination index is not accessible
+ * on current gpu
+ * @param indices_per_source Number of indices supplied for every source in the range
+ * [vertex_input_first, vertex_input_last)
+ * @param global_degree_offset Global degree offset to local adjacency list for every source
+ * represented by current gpu
+ * @return A tuple of device vector containing the sources and destinations gathered locally
+ */
+template <typename GraphViewType, typename EdgeIndexIterator, typename gpu_t>
+std::tuple<rmm::device_uvector<typename GraphViewType::vertex_type>,
+           rmm::device_uvector<typename GraphViewType::vertex_type>,
+           rmm::device_uvector<gpu_t>>
+gather_local_edges(
```

Comment:
Should we place this here or better move to graph_functions.hpp? Most functions under the include/cugraph/prims are the ones that take custom functors and cannot be explicitly instantiated (and we should expose implementation in the header).

I think this function can be explicitly instantiated like other functions in graph_functions.hpp (which have their implementations and explicit instantiation code under cpp/src).

And we may just use int (process ID) for gpu_t (unless we really want to allow users to use 8 bit and 16 bit GPU IDs).